clients/src/test/java/org/apache/kafka/clients/MockClient.java

```diff
@@ -237,10 +237,16 @@ public void send(ClientRequest request, long now) {
                 continue;
 
             AbstractRequest.Builder<?> builder = request.requestBuilder();
-            short version = nodeApiVersions.latestUsableVersion(request.apiKey(), builder.oldestAllowedVersion(),
-                    builder.latestAllowedVersion());
 
             UnsupportedVersionException unsupportedVersionException = null;
+            short version = 0;
+            try {
+                version = nodeApiVersions.latestUsableVersion(request.apiKey(), builder.oldestAllowedVersion(),
+                        builder.latestAllowedVersion());
+            } catch (UnsupportedVersionException e) {
+                unsupportedVersionException = e;
+            }
```

Comment:
It seems that `unsupportedVersionException` will be overwritten in some case as we set it again below. It seems to be that it would be better to structure it as follow:

```
try {
  short version = nodeApiVersions.latestUsableVersion(request.apiKey(), builder.oldestAllowedVersion(),
     builder.latestAllowedVersion());

  UnsupportedVersionException unsupportedVersionException = null;
  if (futureResp.isUnsupportedRequest) {
    unsupportedVersionException = new UnsupportedVersionException(
      "Api " + request.apiKey() + " with version " + version);
  } else {
    AbstractRequest abstractRequest = request.requestBuilder().build(version);
    if (!futureResp.requestMatcher.matches(abstractRequest))
      throw new IllegalStateException("Request matcher did not match next-in-line request "
        + abstractRequest + " with prepared response " + futureResp.responseBody);
  }

  ClientResponse resp = new ClientResponse(request.makeHeader(version), request.callback(), request.destination(),
    request.createdTimeMs(), time.milliseconds(), futureResp.disconnected,
    unsupportedVersionException, null, futureResp.responseBody);
  responses.add(resp);
} catch (UnsupportedVersionException e) {
  ClientResponse resp = new ClientResponse(request.makeHeader(builder.latestAllowedVersion()), request.callback(), request.destination(),
    request.createdTimeMs(), time.milliseconds(), false, unsupportedVersionException, null, null);
  responses.add(resp);
}
```

This would be close to the actual behavior of the regular client. What do you think?

Reply:
makes sense.